### PR TITLE
chore: since/deprecation versioning 

### PR DESCRIPTION
**Pull request by scop**
Closes https://github.com/scop/bash-completion/issues/893
Closes https://github.com/scop/bash-completion/issues/963

### CODE REVIEW

File: bash_completion.d/000_bash_completion_compat.bash [hunk 1]
```diff
@@ -1,23 +1,24 @@
...
+_comp_deprecate_func 2.12 __get_cword_at_cursor_by_ref \
+    _comp__get_cword_at_cursor
+_comp_deprecate_func 2.12 _get_comp_words_by_ref _comp_get_words
+_comp_deprecate_func 2.12 _longopt _comp_longopt
+_comp_deprecate_func 2.12 _split_longopt _comp__split_longopt
+_comp_deprecate_func 2.12 __ltrim_colon_completions _comp_ltrim_colon_completions
 
 # Backwards compatibility for compat completions that use have().
-# @deprecated should no longer be used; generally not needed with dynamically
+# @deprecated TODO should no longer be used; generally not needed with dynamically
```
Comment (scop) [inline bash_completion.d/000_bash_completion_compat.bash:21]: Log diving needed, this was deprecated a long time ago.

File: bash_completion.d/000_bash_completion_compat.bash [hunk 2]
```diff
@@ -106,7 +107,7 @@ _upvar()
 #     current word (default is 0, previous is 1), respecting the exclusions
 #     given at $1.  For example, `_get_cword "=:" 1' returns the word left of
 #     the current word, respecting the exclusions "=:".
-# @deprecated  Use `_comp_get_words cur' instead
+# @deprecated TODO Use `_comp_get_words cur' instead
```
Comment (scop) [inline bash_completion.d/000_bash_completion_compat.bash:110]: Log diving needed, this was effectively deprecated a long time ago.

File: bash_completion.d/000_bash_completion_compat.bash [hunk 3]
```diff
@@ -157,7 +158,7 @@ _get_cword()
 # This is a good alternative to `prev=${COMP_WORDS[COMP_CWORD-1]}' because bash4
 # will properly return the previous word with respect to any given exclusions to
 # COMP_WORDBREAKS.
-# @deprecated  Use `_comp_get_words cur prev' instead
+# @deprecated TODO Use `_comp_get_words cur prev' instead
```
Comment (scop) [inline bash_completion.d/000_bash_completion_compat.bash:161]: Log diving needed, this was effectively deprecated a long time ago.

File: bash_completion.d/000_bash_completion_compat.bash (hunk 1 above)
Comment (scop) [reply bash_completion.d/000_bash_completion_compat.bash]: 20c05b43b, 1.90

File: bash_completion [hunk 4]
```diff
@@ -41,18 +41,24 @@ fi
...
+# @param $2 Old function name
+# @param $3 New function name
+# @since 2.12
 _comp_deprecate_func()
 {
-    if [[ $1 != [a-zA-Z_]*([a-zA-Z_0-9]) ]]; then
-        printf 'bash_completion: %s: %s\n' "$FUNCNAME" "\$1: invalid function name '$1'" >&2
+    if (($# != 3)); then
+        printf 'bash_completion: %s: usage: %s OLD_NAME NEW_NAME DEPRECATION_VERSION\n' "$FUNCNAME" "$FUNCNAME"
+        exit 2
```
Comment (akinomyoga) [inline bash_completion:52]:
```suggestion
        return 2
```

I think `exit` would terminate the shell so we cannot start an interactive session with a mistake.

**Review by akinomyoga — APPROVED** [review]
pre-approve

File: bash_completion (hunk 4 above)
Comment (scop) [reply bash_completion:52]: Oops, indeed, thanks!

**Pull request merged**